File: nightgrass/src/main/java/com/goodfeel/nightgrass/web/CartController.java
package com.goodfeel.nightgrass.web;

import com.goodfeel.nightgrass.dto.CartItemDto;
import com.goodfeel.nightgrass.serviceImpl.CartService;
import com.goodfeel.nightgrass.web.util.CartItemUpdateRequest;
import com.goodfeel.nightgrass.web.util.CartRequest;
import com.goodfeel.nightgrass.web.util.Utility;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.Map;

@Controller
@RequestMapping(path = "/cart")
public class CartController {

    private final Logger logger = LoggerFactory.getLogger(CartController.class);

    @Value("${STRIPE_PUBLIC_KEY}")
    private String stripePublicKey;

    private final CartService cartService;

    public CartController(CartService cartService) {
        this.cartService = cartService;
    }

    // View cart contents
    @GetMapping
    public Mono<String> viewCart(Model model) {
        Flux<CartItemDto> cartItems = cartService.getCartItems();
        model.addAttribute("cartItems", cartItems);
        model.addAttribute("cartTotal", cartService.getTotalPrice());
        model.addAttribute("STRIPE_PUBLIC_KEY", stripePublicKey); // TODO NOT NEEDED HERE?
        return Mono.just("shopping-cart"); // Thymeleaf template for viewing the cart
    }

    @PostMapping(path = "/add")
    public Mono<String> addToCart(@ModelAttribute CartRequest cartRequest, Model model) {
        return cartService.addProductToCart(cartRequest.getProductId())
                .thenReturn("redirect:/product/detail?id=" + cartRequest.getProductId());
    }

    @PostMapping(path = "/remove")
    public Mono<String> removeFromCart(@ModelAttribute CartRequest cartRequest, Model model) {
        return cartService.removeCartItemFromCart(cartRequest.getItemId())
                .then(Mono.fromRunnable(() ->
                        model.addAttribute(
                                "message",
                                "Item removed from the cart successfully!")
                ))
                .thenReturn("redirect:/cart");  // Redirect to the cart view after removal
    }

    @PostMapping("/update-quantity")
    public Mono<ResponseEntity<String>> updateQuantity(@RequestBody CartItemUpdateRequest updateRequest) {
        logger.debug("Received updateRequest request body from front end ajax call: {}", updateRequest);
        return cartService.updateQuantity(updateRequest.getItemId(), updateRequest.getQuantity())
                .map(updated -> ResponseEntity.ok("Quantity updated"))
                .defaultIfEmpty(ResponseEntity.status(HttpStatus.NOT_FOUND).body("Item not found"));
    }

    @PostMapping("/update-total-on-checkbox")
    public Mono<ResponseEntity<Map<String, String>>> updateTotal(@RequestBody CartItemUpdateRequest request) {
        Map<String, String> response = new HashMap<>();
        response.put("message", "Cart total updated on checkbox change");
        return cartService.updateCartTotal(request.getItemId(), request.getIsChecked())
                .thenReturn(ResponseEntity.ok(response));
    }

    @PostMapping("/checkout")
    public Mono<ResponseEntity<Map<String, Long>>> checkout() {
        return Utility.getCurrentUserId().flatMap( userId ->

                cartService.checkout(userId)
                        .map(order -> {
                            Map<String, Long> response = new HashMap<>();
                            response.put("orderId", order.getOrderId());
                            return ResponseEntity.ok(response);
                        })
                        .onErrorResume(e -> {
                            Map<String, Long> errorResponse = new HashMap<>();
                            errorResponse.put("orderId", null);
                            return Mono.just(ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(errorResponse));
                        })
        );
    }

    @GetMapping("/total")
    public Mono<ResponseEntity<BigDecimal>> getCartTotal() {
        return cartService.getTotalPrice()
                .map(ResponseEntity::ok)
                .defaultIfEmpty(ResponseEntity.status(HttpStatus.NOT_FOUND).body(BigDecimal.ZERO));
    }

}

File: nightgrass/src/main/kotlin/com/goodfeel/nightgrass/web/CartController.kt
package com.goodfeel.nightgrass.web

import com.goodfeel.nightgrass.data.Order
import com.goodfeel.nightgrass.serviceImpl.CartService
import com.goodfeel.nightgrass.web.util.CartItemUpdateRequest
import com.goodfeel.nightgrass.web.util.AddCartRequest
import com.goodfeel.nightgrass.web.util.Utility
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.*
import reactor.core.publisher.Mono
import java.math.BigDecimal

@Controller
@RequestMapping("/cart")
class CartController(private val cartService: CartService) {
    private val logger: Logger = LoggerFactory.getLogger(CartController::class.java)

    @Value("\${STRIPE_PUBLIC_KEY}")
    private val stripePublicKey: String? = null

    // View cart contents
    @GetMapping
    fun viewCart(model: Model): Mono<String> {
        val cartItemsFlux = cartService.getCartItems()
        val totalPriceMono = cartService.getTotalPrice()

        return Mono.zip(cartItemsFlux.collectList(), totalPriceMono)
            .map { tuple ->
                val cartItems = tuple.t1
                val cartTotal = tuple.t2
                model.addAttribute("cartItems", cartItems)
                model.addAttribute("cartTotal", cartTotal)
                "shopping-cart"
            }
    }

    @PostMapping("/add")
    fun addToCart(@ModelAttribute addCartRequest: AddCartRequest): Mono<String> {
        logger.debug("Adding product ${addCartRequest.productId} with properties: ${addCartRequest.itemId}")
        return cartService.addProductToCart(addCartRequest)
            .thenReturn("redirect:/product/detail?id=${addCartRequest.productId}")
    }

    @PostMapping("/remove")
    fun removeFromCart(@ModelAttribute addCartRequest: AddCartRequest, model: Model): Mono<String> {
        return addCartRequest.itemId?.let { itemId ->
            cartService.removeCartItemFromCart(itemId)
                .doOnSuccess {
                    model.addAttribute("message", "Item removed from the cart successfully!")
                    logger.debug("Cart item $itemId removed successfully")
                }
                .onErrorResume { error ->
                    logger.error("Error removing item $itemId: ${error.message}")
                    model.addAttribute("errorMessage",
                        "Failed to remove the item from the cart.")
                    Mono.empty()
                }
                .thenReturn("redirect:/cart")
        } ?: run {
            logger.error("Invalid itemId in request") // Logging null case
            Mono.just("redirect:/error")
        }
    }

    @PostMapping("/update-quantity")
    fun updateQuantity(@RequestBody updateRequest: CartItemUpdateRequest): Mono<ResponseEntity<String>> {
        logger.debug("Received updateRequest request body from front end ajax call: {}", updateRequest)
        return cartService.updateQuantity(updateRequest.itemId, updateRequest.quantity!!) // TODO
            .map { ResponseEntity.ok("Quantity updated") }
            .defaultIfEmpty(ResponseEntity.status(HttpStatus.NOT_FOUND).body("Item not found"))
            .onErrorResume { ex ->
                logger.error("Error occurred while updating quantity", ex)
                Mono.just(ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Failed to update quantity"))
            }
    }

    @PostMapping("/update-total-on-checkbox")
    fun updateTotal(@RequestBody request: CartItemUpdateRequest): Mono<ResponseEntity<Map<String, String>>> {
        val response: MutableMap<String, String> = HashMap()
        response["message"] = "Cart total updated on checkbox change"
        return cartService.updateCartTotal(request.itemId, request.isChecked!!) // TODO
            .thenReturn(ResponseEntity.ok(response))
    }

    /**
     * When customer press Checkout button on shopping cart page, front end javascript post a request
     * to create order, insert order items and clean up shopping cart.
     * @return
     */
    @PostMapping("/checkout")
    fun checkout(): Mono<ResponseEntity<Map<String, Long?>>> {
        return Utility.currentUserId.flatMap { userId: String ->
            cartService.createOrderAndCleanupShoppingCart(userId)
                .map { order: Order ->
                    ResponseEntity.ok(mapOf("orderId" to order.orderId))
                }
                .onErrorResume { e ->
                    logger.error("Error occurs on shopping cart checkout: $e")
                    Mono.just(
                        ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                            .body(mapOf("orderId" to null))
                    )
                }
        }
    }

    @get:GetMapping("/total")
    val cartTotal: Mono<ResponseEntity<BigDecimal>>
        get() = cartService.getTotalPrice()
            .map<ResponseEntity<BigDecimal>> { body: BigDecimal? -> ResponseEntity.ok(body) }
            .defaultIfEmpty(ResponseEntity.status(HttpStatus.NOT_FOUND).body(BigDecimal.ZERO))
}
